Add note about set_dir() and metadata (#3282)
quiltdata/quilt: docs/Troubleshooting.md

1. Send screenshots to [Quilt support](mailto:support@quiltdata.io).
1. It is not recommended that you adjust ElasticSearch via Edit domain, as these
changes will be lost the next time that you update Quilt

> ## Missing metadata when working with data packages via the API

Reviewer: Maybe "Quilt packages" or just "packages"?

> `Package.set_dir()` on the package root (".") overrides package-level metadata.
> If you do not provide `set_dir(".", foo, meta=baz)` with a value for `meta=`,
> `set_dir` will set package-level metadata to `None`.

A common pattern is to `Package.browse()` to get the most recent
version of a package, and then `Package.push()` updates.
You can preserve package-level metadata when calling `set_dir(".", ...)`
as follows:

<!--pytest.mark.skip-->
```python
import quilt3

p = quilt3.Package.browse(
    f"user-packages/{package_name}", 
    f"s3://{bucket}"
)

p.set_dir(
    ".",

>     f"s3://{bucket}/user-packages/{package_name}/",

Reviewer 2: Wait the right hand side of this setdir is super duper confusing and wrong. Use “s3://somewhere/else” (not literally that but you get the idea).

Reviewer 2: I’d recommend much simpler strings in general as well

1. Send screenshots to [Quilt support](mailto:support@quiltdata.io).
1. It is not recommended that you adjust ElasticSearch via Edit domain, as these
changes will be lost the next time that you update Quilt

## Missing metadata when working with data packages via the API

> `Package.set_dir()` on the package root (".") overrides package-level metadata.
> If you do not provide `set_dir(".", foo, meta=baz)` with a value for `meta=`,
> `set_dir` will set package-level metadata to `None`.

> > This is because _folder-level_ metadata overrides _package-level_ metadata.

Reviewer 2: Eliminate this sentence

1. Send screenshots to [Quilt support](mailto:support@quiltdata.io).
1. It is not recommended that you adjust ElasticSearch via Edit domain, as these
changes will be lost the next time that you update Quilt

## Missing metadata when working with data packages via the API

> If you are browsing and pushing a package using the Python `quilt3`
> API be careful to note that `Package.set_dir()` on the root path
> (`.`) currently sets metadata to `None`:

Reviewer 2: ```suggestion
> `Package.set_dir()` on the package root (".") overrides package-level metadata.
> If you do not provide `set_dir(".", foo, meta=baz)` with `meta=`, package-level
> metadata will be set to `None`.
```

> This is because _folder-level_ metadata overrides _package-level_ metadata.

A common development pattern is to `Package.browse()` to get the most recent
version of a package, and then `Package.push()` updates. Be sure
to add any **existing package metadata** to ensure the metadata is not replaced:

<!--pytest.mark.skip-->
```python
import quilt3

> p = quilt3.Package()
> p.browse(
>     name=f"user-packages/{package_name}", 
>     registry=f"s3://{bucket}"
> )
> 
> # Get existing package-level metadata
> metadata = p.meta
> 
> # Add all files from path to the package and preserve existing metadata
> p.set_dir(
>     lkey=".",
>     path=f"s3://{bucket}/user-packages/{package_name}/",
>     meta=metadata
> )

Reviewer 2: No reason to do p = q3.Package. That line has no effect whatsoever as written.
Simpler not to have an intermediate variable as well.
```suggestion
p = quilt3.Package.browse(
    name=f"user-packages/{package_name}", 
    registry=f"s3://{bucket}"
)

# Add all files from path to the package and preserve existing metadata
p.set_dir(
    lkey=".",
    path=f"s3://{bucket}/user-packages/{package_name}/",
    meta=p.meta
)
```

# Push changes to the S3 registry

> p.push(
>     name=f"user-packages/{package_name}",
>     registry=f"s3://{bucket}",
>     message=f"Updating package {package_name}",
>     workflow="default"
> )

Reviewer 2: I'd say no reason to mention workflows here; also naming the default seems odd since it's default.
```suggestion
p.push(
    name=f"user-packages/{package_name}",
    registry=f"s3://{bucket}",
    message=f"Updating package {package_name}"
)
```
